src/cumulus/internal.clj

```diff
@@ -0,0 +1,302 @@
+(ns cumulus.internal
+  (:require
+    [clojure.string :as str]))
+
+
+(defn reqd
+  "Get required parameter k from map m. Throw exception if not found."
+  [m k]
+  (if (contains? m k)
+    (name (get m k))
+    ;(get m k)
+    (throw (IllegalArgumentException.
+             (format "Key %s not found in %s" (pr-str k) (pr-str m))))))
+
+
+(defn expected
+  "Throw illegal input exception citing `expectation` and what was `found` did not match. Optionally accept a predicate
+  fn to test `found` before throwing the exception."
+  ([expectation found]
+    (throw (IllegalArgumentException.
+             (format "Expected %s, but found (%s) %s" expectation (class found) (pr-str found)))))
+  ([pred expectation found]
+    (when-not (pred found)
+      (expected expectation found))))
+
+
+(defn as-str
+  "Convert `x` to string and return it"
+  [x]
+  (if (or (keyword? x) (symbol? x))
+    (name x)
+    (str x)))
+
+
+(defn type-check
+  [db-params key]
+  (if (not (integer? (get db-params key)))
+    ((throw (IllegalArgumentException. (format "Expected Integer but found %s" (pr-str (get db-params key))))))))
+
+
+(defn type-check-string
+  [db-params key]
+  (if (not (or (string? (get db-params key)) (keyword? (get db-params key)) (symbol? (get db-params key))))
+    ((throw (IllegalArgumentException. (format "Expected String but found %s" (pr-str (get db-params key))))))))
+
+
+(defn parse-fn
+  [db-params key]
+  (try 
+    (Integer/parseInt (get db-params key))
+    (catch NumberFormatException _
+      (throw (IllegalArgumentException.(format "Expected Integer string but found %s" (get db-params key)))))))
+
+
+;; ===== Helper Functions =====
+
+
+(defn raw-params
+  [driver-classname
+   jdbc-url
+   test-query]
+  {:classname  driver-classname
+   :jdbc-url   jdbc-url
+   :test-query test-query})
+
+
+(defn R 
+  [m k]
+  (reqd m k))
+
+
+(defn P 
+  ([m k]
+    (get m k))
+  ([m k d]
+    (get m k d)))
+
+
+(defn Q 
+  ([m k]
+    (if-let [v (P m k)]
+      (str ":" v)
+      ""))
+  ([m k d]
+    (str ":" (P m k d))))
+
+
+(defn jdbc
+  [m]
+  (merge m (raw-params
+             (format "%s" (R m :classname))
```

Comment:
Consider omitting `format` as it is serving no purpose here. Rather, ensure that the driver class-name is a string before returning the value.